### PR TITLE
Add dark mode

### PR DESCRIPTION
**Pull request by devnikesh**
## **What**  
I've added support for Dark Mode according to theme preference. It show's up according to the user theme preference in their browser settings. For more background, see ticket [**Add Dark Mode #11**](https://github.com/darekkay/static-marks/issues/11) **Repo:** [darekkay/static-marks](https://github.com/darekkay/static-marks)

## **Why**  
These changes help the users to reduce  reduced glare and blue light exposure and extend battery and was in the issue section `label:Enhancement` 

## **How**  
In App.jsx, we’re checking if the user has set a theme preference in their browser settings. Then addressing them according to the preference by providing theme data to App, which then gets styling through App.scss for Dark and Light mode.

## **Testing**  
 While testing the `window.matchMedia()` method was throwing errors `TypeError: window.matchMedia is not a function` , So I had to add a `matchMedia.js` mock file to pass the Jest Test.

**Links Used to solve the test issue**
* [Mocking methods which are not implemented in JSDOM](https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom)
* [Cannot read property ‘matches’ of undefined](https://github.com/facebook/create-react-app/issues/10126)

## **Screenshots**  

**Test Status**: Passed
Screenshot (1352)

**Dark Mode**
Dark Theme

## **Anything Else?**  
As mentioned in the [**Add Dark Mode #11**](https://github.com/darekkay/static-marks/issues/11) issue, there's also a requirement for Add theme switcher, so we will be adding that functionality in coming PR.
.

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -38,9 +38,12 @@ const Collections = ({ projects, projectKey, title, currentFilter }) => {
 const App = (props) => {
   const { title, projectKeys, projects, initialFilter } = props;
   const [currentFilter, applyFilter] = useState(initialFilter);
+  const defaultWhite = window.matchMedia(
+    "(prefers-color-scheme: white)"
+  ).matches;
```
Comment (darekkay) [inline src/App.jsx]: NICE: This code can be extracted out of `App` to prevent the recalculation on every rerender.

File: src/App.jsx [hunk 2]
```diff
@@ -38,9 +38,12 @@ const Collections = ({ projects, projectKey, title, currentFilter }) => {
 const App = (props) => {
   const { title, projectKeys, projects, initialFilter } = props;
   const [currentFilter, applyFilter] = useState(initialFilter);
+  const defaultWhite = window.matchMedia(
+    "(prefers-color-scheme: white)"
```
Comment (darekkay) [inline src/App.jsx]:
```suggestion
    "(prefers-color-scheme: light)"
```

See [specs](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme)

Right now, your code does not make it possible to use the light theme. However, after the proposal above, the dark mode is not working anymore:

image

Additionally, the current code uses dark mode by default if the use has no preference at all, which I personally don't like as a default value. We should probably invert the logic.

**Review by darekkay — CHANGES_REQUESTED** [review]
Thanks @devnikesh , this looks like a good work in progress. However, there are some open tasks left:

- Light mode is currently broken, even if the user prefers light mode
- Some parts of the applications still use "light" mode in dark mode (main background, collection names, bucket titles, filter input)
- There are some linter issues, see GitHub action results

**Comment by devnikesh** [comment]
@darekkay I apologize for that, I'm on it, the light and dark mode logic is solved on my end, I'm trying to fix the unapplied dark theme on (main background, collection names, bucket titles, filter input)

**Comment by darekkay** [comment]
@devnikesh No need to apologize! As mentioned in the main issue, you can probably solve most of this by using the [new atomic color classes](https://github.com/darekkay/static-marks-app/blob/master/src/styles/_atomic.scss) that will match the CSS custom properties. In other words, instead of adjusting CSS and using `var(...)` directly, you can often use classes like `.text-default`, `.text-offset`, `.bg-offset` etc. instead.

**Comment by devnikesh** [comment]
@darekkay  yeah sure, will implement this in the next PR, Thank you.

**Comment by devnikesh** [comment]
@darekkay  yeah sure, will implement this, Thank you.

**Comment by devnikesh** [comment]
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned, hope it works. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

Dark Theme Enabled

**Comment by devnikesh** [comment]
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

Dark Theme Enabled

**Comment by devnikesh** [comment]
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons, Footer titles)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

Dark Theme Enabled

**Comment by devnikesh** [comment]
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons, Footer titles)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

Dark Theme enabled

**Comment by devnikesh** [comment]
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons, Footer titles)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

Dark Theme enabled

**Comment by darekkay** [comment]
Thanks @devnikesh ! I'll merge your changes into a feature branch `dark-mode`, so I can do some further alignments with some of my recent work on color schemes.

**Review by darekkay — APPROVED** [review]

**Pull request merged**